opsdroid/skill.py

```diff
@@ -0,0 +1,37 @@
+"""Base class for class-based skills
+"""
+from opsdroid.matchers import add_skill_attributes
+
+class Skill:
+    def __init__(self, opsdroid, config, *args, **kwargs):
+        super(Skill, self).__init__()
+
+        self.opsdroid = opsdroid
+        self.config = config
+
+        for attr in self.__dir__():
+            try:
+                method = getattr(self, attr)
+            except Exception:
+                continue
+
+            for decor_params in getattr(method, '__matchers__', []):
+                decor_func, args, kwargs = decor_params
+
+                method.__skill_attrs__['config'] = self.config
+                decorator = decor_func(*args, **kwargs)
+                setattr(self, attr, decorator(method))
+
+    @classmethod
+    def set_matcher(cls, decor_func, *args, **kwargs):
```

Comment:
Could you explain why you have taken this approach?

I had imagined you would be able to still use the regular decorators on the methods. It may require updating the decorators but I would prefer that to forking the api.

Reply:
If you decorate a class method the decorated function is a class method (not bound to an instance yet). But when you activate a skill, you have to work on an object instance, not the unbound class method.

So either you store the skill instance(s) and for each unbound method call all the relevant instance methods, or do it this way :(

Reply:
I give it a go, but i’m afraid that requires bigger changes where we actually call the skill functions. We shall see.

Reply:
Now that i think of it, that might require a huge overhaul of parsers, including external ones (ie. not written by opsdroid maintainers). But that might also apply to this solution, so i definitely give it a go now.

Reply:
There are no external parsers. They are all core modules.

Reply:
Is it not possible to add external parsers, like it is with skills, connectors, and databases? If not, i can do the modifications, of course.

Reply:
Not currently.